### PR TITLE
Linkage Monitor to pass system properties when building a POM

### LINKED ISSUE #923: Linkage Monitor cannot determine Java version for profile activation
https://github.com/googleapis/google-cloud-java/pull/6306 failed:

```
[ERROR] Failed to determine Java version for profile checkstyle-tests @ com.google.cloud:google-cloud-shared-config:0.1.2
	at org.apache.maven.model.building.DefaultModelProblemCollector.newModelBuildingException(DefaultModelProblemCollector.java:197)
	at org.apache.maven.model.building.DefaultModelBuilder.build(DefaultModelBuilder.java:490)
	at org.apache.maven.model.building.DefaultModelBuilder.build(DefaultModelBuilder.java:432)
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.buildModelWithSnapshotBom(LinkageMonitor.java:234)
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.copyWithSnapshot(LinkageMonitor.java:248)
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.run(LinkageMonitor.java:94)
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.main(LinkageMonitor.java:82)

```

There's a profile that should be activated for Java 8: https://github.com/googleapis/java-shared-config/blame/master/pom.xml#L368-L408 . But it seems Linkage Monitor cannot determine Java version. We have similar issue in past for the enforcer rule, where  "java.version" system property was missing: https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/817#issuecomment-516970945 .

TODO
- [x] Try to reproduce the error in local Linkage Monitor invocation
  - [x] locally install chingor13's core-pom branch
  - [x] Run Linkage Monitor. Does the error come out?
    Yes.
- [x] Determine where DefaultModelBuilder trying to read the java version for profile activation. JdkVersionProfileActivator (screenshot below).
- [x] Add the "java.version" system property to appropriate sources.
  We can target for Java 8 for now.

SystemProperty was empty when checking JDK version.
image

DefaultModelBuilder.build( ModelBuildingRequest request ) has

```
        DefaultProfileActivationContext profileActivationContext = getProfileActivationContext( request );
```

This `request` is from `ModelBuildingRequest request`.

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #923

### CODE REVIEW

**Comment by suztomo** [comment]
```
[ERROR] testBuildModelWithSnapshotBom_JdkVersionActivation(com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitorTest)  Time elapsed: 0.017 s  <<< ERROR!
org.apache.maven.model.building.ModelBuildingException: 
1 problem was encountered while building the effective model for com.google.cloud:google-cloud-core-parent:1.91.0
[FATAL] Non-resolvable parent POM for com.google.cloud:google-cloud-core-parent:1.91.0: Could not find artifact com.google.cloud:google-cloud-shared-config:pom:0.1.2 and 'parent.relativePath' points at wrong local POM @ 

```

**Comment by suztomo** [comment]
9fed3ed to avoid using google-cloud-core-parent worked.

**Comment by suztomo** [comment]
80c2427 to add Maven Central to model building request worked.

**Comment by elharo** [comment]
FYI, I'm on vacation until Wednesday and busier than I usually am when
I'm not on vacation. :-) Perhaps David or Garrett can review.

On Fri, Sep 20, 2019 at 1:23 PM Tomo Suzuki <notifications@github.com> wrote:
>
> @suztomo requested your review on: #924 Linkage Monitor to pass system properties when building a POM.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
Elliotte Rusty Harold
elharo@macfaq.com

**Comment by suztomo** [comment]
Thank you for heads-up.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 1]
```diff
@@ -206,14 +207,19 @@ static Model buildModelWithSnapshotBom(
...
+            null,
+            repositorySystem,
+            new DefaultRemoteRepositoryManager(),
+            ImmutableList.of(CENTRAL), // Needed when parent pom is not locally available
+            null,
+            null));
+    // Profile activation needs JDK version through system properties
+    // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/923
+    Properties modelBuildingProperties = modelRequest.getSystemProperties();
+    modelBuildingProperties.putAll(System.getProperties());
```
Comment (suztomo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: testBuildModelWithSnapshotBom_JdkVersionActivation fails without this line due to the same error as #923

Comment (netdpb) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: This allows the system properties to override the default request properties. Is that what we want?

Comment (suztomo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: Yes. As discussed it's updated to call `setSystemProperties`.

**Pull request merged**